### PR TITLE
Store meta-data in observations as JSON

### PR DESCRIPTION
**Pull request by tomaszmrugalski**
This is definitely a WIP, but the overall goal is to be able to store meta-data with the observations. By meta-data, I mean many aspects. Some examples are:

- SDR used
- LNA, filter being used
- what kind fo antenna is used
- what kind of recipe is used
- etc.

### CODE REVIEW

**Comment by tomaszmrugalski** [comment]
That's how the observations list looks like now. Two extra columns were added: frequency and protocol. Many more are stored in the config field as JSON.

Screenshot 2021-11-27 at 00 59 33

All available parameters are printed in the station and observation views.

**Comment by tomaszmrugalski** [comment]
This is probably enough for an early review. In particular, I'm interested if this approach is OK or not. If it is, I'll go ahead and modify the submit_obs code to provide parameters for new observations.

The upgrade script retroactively sets the configuration parameters (frequency, hardware used etc).

File: server/db/svarog-11.psql [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+-- Now we need to migrate existing data. This is really specific to station 1.
+-- Let's add protocol transmission. So far we were able to receive APT and LRPT.
+update observations set config = config || '{"proto": "APT" }' where sat_id in (25338, 28654, 33591) AND station_id = 1;
+update observations set config = config || '{"proto": "LRPT" }' where sat_id in (40069, 44387) AND station_id = 1;
+
+-- All transmissions received so far were done using crossed dipole
+update observations set config = config || '{"antenna": "WiMo TA-1", "antenna-type": "crossed dipole" }' WHERE station_id = 1;
+
+-- Add frequencies. We haven't done any Doppler magic, so everything was received on base frequencies.
+update observations set config = config || '{"freq": 137.62 }' WHERE sat_id = 25338 and station_id = 1; -- NOAA 15
```
Comment (fivitti) [inline server/db/svarog-11.psql:15]: It should be "frequency". If we will promote the abbreviation then one user will use "freq" and another "f".

File: server/db/svarog-11.psql [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+update observations set config = config || '{"freq": 137.912 }' WHERE sat_id = 28654 and station_id = 1; -- NOAA 18
+update observations set config = config || '{"freq": 137.1 }' WHERE sat_id = 33591 and station_id = 1; -- NOAA 19
+update observations set config = config || '{"freq": 137.1 }' WHERE sat_id = 40069 and station_id = 1; -- METEOR M-2
+
+-- Now set the SDR being used. I don't remember the exact date when AirSpy was installed. However, shortly afterwards
+-- the PR for adding generic RX recipe (PR#50) was merged. The merge was on 2021-06-09, so let's assume
+-- that's the date of AirSpy upgrade.
+
+UPDATE observations set config = config || '{"sdr": "RTL-SDRv3"}' WHERE station_id = 1 and aos<'2021-06-09'::date;
```
Comment (fivitti) [inline server/db/svarog-11.psql:24]: Use "receiver" instead of "sdr". Some stations can use a dedicated devices to receive the radio signal.

File: server/db/svarog-11.psql [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+UPDATE observations set config = config || '{"sdr": "AirSpy Mini"}' WHERE station_id = 1 and aos>='2021-06-09'::date;
+
+-- The last METEOR observation recorded is from 2021-03-26. That's because the newly installed LNA + filter
+-- (Nooelec NOAA) has a filter that is wide enough for NOAA, but too narrow for Meteor.
+UPDATE observations SET config = config || '{"lna":"Nooelec NOAA", "filter":"Nooelec NOAA"}'
```
Comment (fivitti) [inline server/db/svarog-11.psql:29]: Do we need the "lna" and "filter" both? It looks like duplicated data.

File: server/templates/obslist.html [hunk 4]
```diff
@@ -68,6 +68,8 @@ <h3>Observations</h3>
         <th class="text-center"><abbr title="Two Line Element">TLE</abbr></th>
         <th class="text-center">Rating</th>
         <th>Station</th>
+        <th>Frequency</th>
+        <th>Protocol</th>
```
Comment (fivitti) [inline server/templates/obslist.html:72]: I think that the protocol shouldn't be visible on the main list.

- The protocol Is little related to the frequency because the satellites use one protocol per frequency.
- The protocol may be unknown, but the frequency always is known.
- User may be unsure what protocol is used. For example, when the embedded metadata is received.

File: server/templates/obslist.html [hunk 5]
```diff
@@ -103,6 +103,17 @@ <h3>Observations</h3>
                     {{ obs.station_name }}
                 </a>
             </td>
+            <td>
+                {% if obs.config.freq is defined %}
+                {{ "%.3f MHz"|format(obs.config.freq) }}
```
Comment (fivitti) [inline server/templates/obslist.html:108]: Could you document that we expect the frequency in megahertz? Maybe, we should use just the hertz and use a template filter to convert the unit? I think that it will be the simplest and best solution.

File: server/templates/station.html [hunk 6]
```diff
@@ -26,7 +26,14 @@ <h3>Ground Station  {{ station.name }}</h3>
         <b>About the station</b>: {{ station.descr }}
     </p>
     <p>
-        <b>Current configuration:</b> {{ station.config }}
+        <b>Current configuration:</b>
+        <ul>
+        {% for c in station.config %}
+        {% if station.config[c] %}
```
Comment (fivitti) [inline server/templates/station.html:32]: Use the `is defined` statement. Now, the value that equals `0` or `""` will not be visible.

File: server/tests/test_repository_postgres.py [hunk 7]
```diff
@@ -50,6 +50,10 @@ def test_db_version(self, repository):
         migration_numbers = []
         for filename in os.listdir(directory):
             base, _ = os.path.splitext(filename)
+
+            # Skip any files that do not conform to the upgrade script naming convention
+            if base[:7] != "svarog-":
```
Comment (fivitti) [inline server/tests/test_repository_postgres.py:55]: Use `.startswith` method of string.

**Review by fivitti — CHANGES_REQUESTED** [review]

File: server/db/svarog-11.psql (hunk 3 above)
Comment (tomaszmrugalski) [reply server/db/svarog-11.psql:29]: Right now we have one device that has two features integrated, but I also have separate filter and separate LNA. Those can be different.

File: server/templates/obslist.html (hunk 4 above)
Comment (tomaszmrugalski) [reply server/templates/obslist.html:72]: I can easily hide it, but I was thinking that it's useful to compare APT to LRPT for example. People can draw conclusions from that. I was thinking that long term we would expose all of them, so you could filter and compare various scenarios, APT vs LRPT, reception with and without LNA etc. Also, having the protocol clickable (link to wikipedia) would have some educational value. Finally, the third argument is about emphasizing new protocols, which I very much hope we'll see soon.

So, does that sound convincing? If not, I'll remove it.

File: server/tests/test_repository_postgres.py (hunk 7 above)
Comment (tomaszmrugalski) [reply server/tests/test_repository_postgres.py:55]: Updated as suggested.

File: server/db/svarog-11.psql (hunk 2 above)
Comment (tomaszmrugalski) [reply server/db/svarog-11.psql:24]: Renamed as suggested.

File: server/templates/station.html (hunk 6 above)
Comment (tomaszmrugalski) [reply server/templates/station.html:32]: Added as suggested.

File: server/templates/obslist.html (hunk 5 above)
Comment (tomaszmrugalski) [reply server/templates/obslist.html:108]: The docs now say it's in MHz. If you want, I can change it to Hz.

Comment (tomaszmrugalski) [reply server/templates/obslist.html]: After second though (and looking at what we use in the station's code), I agree that using Hz is the way to go. The observations list looks the same as before, but the underlying data is in Hz. You can see that when looking at specific observation page.

**Comment by tomaszmrugalski** [comment]
There's a handful of unrelated changes that were discovered or needed for other reasons:
- added setup and teardown functions to repository unit-tests. Don't do anything yet, but will be needed for extended tests
- corrected whitespace and some descriptions in pagination.py
- fixed exception on stations list if station didn't have registered field set
- updated schema, so registered field defaults to now() if not explicitly specified

I hope those are small enough. If you feel any of them should go to a separate PR, let me know.

**Comment by tomaszmrugalski** [comment]
This is a request to do a preliminary review. I'm specifically interested if the ability to pass meta-data (called 'config', but perhaps calling it metadata would be more understandable?) is reasonable. I understand that there are multiple tests and comments missing. The feedback I'm looking for is for the concept.

There's a lot of changes:
- if run as stand-alone script, prints details on stdout
- handles various problems better, including printing response in case of failure
- MIME type is now generic, not limited to png
- one test added for submitobs (many still missing)
- support for png, jpg, gif, txt, csv, json files
- support for uploading data types that are NOT images (so thumbnail can't be generated)
- server logs more data during observation reception
- (unrelated) server no longer memory leaks when generating pass plots
- refactored ALLOWED_FILE_TYPES (because multiple extensions could lead to text/plain type)
- removed unused notes field
- added config field (technically it's a string, but should contain JSON. This is checked in the submitobs.py)

The bottom line is that you can now run the following command:
```python3 submitobs.py image1.png,image6.json "NOAA 18" "2021-11-27 21:35" "2021-11-27 21:37" "2021-11-27 21:39" '{"antenna":"corner", "receiver":"RTL-SDRv3", "frequency": 137100000 }' 0.3```

and it will upload multiple files (and possibly logs), will put the metadata into the observation and will generally work.

File: server/db/svarog-11.psql (hunk 3 above)
Comment (fivitti) [reply server/db/svarog-11.psql:29]: Ok, but the configuration isn't a fixed list, and you can omit the fields that aren't necessary. We use only the LNA now then we shouldn't add the filter field. 
If another station will have only a filter, it omits the LNA.

File: server/templates/obslist.html (hunk 4 above)
Comment (fivitti) [reply server/templates/obslist.html:72]: Ok, it makes sense. But we should be careful that the table isn't too wide.

File: server/app/controllers/receive.py [hunk 8]
```diff
@@ -42,17 +42,28 @@ class RequestArguments(TypedDict):
...
+    ".json": "application/json",
+    ".txt": "text/plain",
+    ".log": "text/plain",
+    ".csv": "text/plain"
 }
 
 get_extension = lambda p: os.path.splitext(p)[1]
-is_allowed_file = lambda f: f.mimetype in ALLOWED_FILE_TYPES and \
-                            ALLOWED_FILE_TYPES[f.mimetype] == get_extension(f.filename)
+is_allowed_file = lambda f: get_extension(f.filename) in ALLOWED_FILE_TYPES and \
```
Comment (fivitti) [inline server/app/controllers/receive.py:65]: Use: `ALLOWED_FILE_TYPES.get(get_extension(f.filename)) == f.mimetype)`

File: server/app/controllers/receive.py [hunk 9]
```diff
@@ -104,19 +116,21 @@ def receive(station_id: str, args: RequestArguments):
...
     # Select thumbnail source file
     thumbnail_source_entry = first(lambda f: f[1].mimetype.startswith("image/"), file_entries)
     if thumbnail_source_entry is None:
-        abort(400, description="Missing imagery file")
-        return
-
-    thumb_source_filename, _ = thumbnail_source_entry
-    thumb_filename = "thumb-%s-%s" % (str(uuid.uuid4()), thumb_source_filename)
+        app.logger.info(f"No suitable images for thumbnail, using the default image {DEFAULT_THUMBNAIL}")
+        thumb_filename = DEFAULT_THUMBNAIL
```
Comment (fivitti) [inline server/app/controllers/receive.py:130]: Don't use a dedicated constant, just a `None` value.

File: server/app/controllers/receive.py [hunk 10]
```diff
@@ -173,10 +187,12 @@ def receive(station_id: str, args: RequestArguments):
             app.logger.error("Failed to write %s (image_root=%s): %s" % (path, root, e))
             return abort(503, "Unable to write file %s. Disk operation error." % filename)
 
-    # Make thumbnail
-    thumb_source_path = os.path.join(root, thumb_source_filename)
-    thumb_path = os.path.join(root, "thumbs", thumb_filename)
-    make_thumbnail(thumb_source_path, thumb_path)
+    if (thumb_filename != DEFAULT_THUMBNAIL):
```
Comment (fivitti) [inline server/app/controllers/receive.py:190]: Remove the brackets.

File: server/db/svarog-11.psql [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+-- Now we need to migrate existing data. This is really specific to station 1.
+-- Let's add protocol transmission. So far we were able to receive APT and LRPT.
+UPDATE observations SET config = config || '{"protocol": "APT" }' WHERE sat_id IN (25338, 28654, 33591) AND station_id = 1;
+UPDATE observations SET config = config || '{"protocol": "LRPT" }' WHERE sat_id IN (40069, 44387) AND station_id = 1;
+
+-- All transmissions received so far were done using crossed dipole
+UPDATE observations SET config = config || '{"antenna": "WiMo TA-1", "antenna-type": "crossed dipole" }' WHERE station_id = 1;
+
+-- Add frequencies. We haven't done any Doppler magic, so everything was received on base frequencies.
+UPDATE observations SET config = config || '{"frequency": 137.62e6 }' WHERE sat_id = 25338 AND station_id = 1; -- NOAA 15
```
Comment (fivitti) [inline server/db/svarog-11.psql:15]: Frequency is in Hz or MHz, finally?

File: server/templates/obslist.html [hunk 12]
```diff
@@ -103,6 +103,17 @@ <h3>Observations</h3>
                     {{ obs.station_name }}
                 </a>
             </td>
+            <td>
+                {% if obs.config.frequency is defined %}
+                {{ "%.3f MHz"|format(obs.config.frequency/1000000) }}
```
Comment (fivitti) [inline server/templates/obslist.html:108]: We will need a special frequency helper here (not necessarily now). The frequencies below 1MHz should be formatted in Hz, above 1GHz in GHz.

File: server/tests/test_repository.py [hunk 13]
```diff
@@ -5,6 +5,14 @@
 
 class RepositoryTests(unittest.TestCase):
 
+    def setUp(self):
```
Comment (fivitti) [inline server/tests/test_repository.py:8]: It should be added only if necessary.

File: doc/submit_obs.md [hunk 14]
```diff
@@ -0,0 +1,21 @@
+# Submitting observations
+
+This page describes the protocol and parameters necessary for submitting new
+observations. It is a work in progress.
+
+## Configuration
+
+Configuration should be submitted as JSON. There's a number of parameters that
+are (almost) mandatory:
```
Comment (fivitti) [inline doc/submit_obs.md:9]: Maybe we should print the warnings when the "almost" mandatory field is missing?

File: station/receiver.py [hunk 15]
```diff
@@ -89,9 +89,10 @@ def cmd():
             rating = get_rating_for_product(product[1], satellite.get("rate"))
             logging.info("Product %s got rating %s" % (product[1], rating))
             # TODO: Submit ALL products and logs
+            files = [ product[1] ]
```
Comment (fivitti) [inline station/receiver.py:92]: Maybe, we should send all products?

File: station/submitobs.py [hunk 16]
```diff
@@ -56,29 +62,81 @@ def get_tle(sat_name: str, date: datetime.datetime) -> Optional[List[str]]:
...
+        "log": "text/plain",
+        "png": "image/png",
+        "svg": "image/svg+xml",
+        "txt": "text/plain",
+        "wav": "audio/wav"
+    }
+
+    txt = filename.lower()
+    offset = txt.rfind('.')
+    ext = txt[offset+1:] if offset != -1 else ""
```
Comment (fivitti) [inline station/submitobs.py:95]: Never ever Use `os.path.splitext` function.

File: station/submitobs.py [hunk 17]
```diff
@@ -56,29 +62,81 @@ def get_tle(sat_name: str, date: datetime.datetime) -> Optional[List[str]]:
...
+    return known_types[ext] if ext in known_types else "application/octet-stream"
```
Comment (fivitti) [inline station/submitobs.py:96]: Better: `known_types.get(ext, "application/octet-stream")`

File: station/submitobs.py [hunk 18]
```diff
@@ -56,29 +62,81 @@ def get_tle(sat_name: str, date: datetime.datetime) -> Optional[List[str]]:
...
         form_data["tle"] = tle
 
-    file_obj = open(data.image_path, 'rb')
+    # Now process the files
+    files = {}
+    cnt = 0
     body: Dict[str, Any] = {
-        "file": file_obj
     }
+    for f in data.image_path:
```
Comment (fivitti) [inline station/submitobs.py:127]: `image_path` is a singular verb. But it's a list of paths now. Also, these paths mustn't be image paths.
In this context (reading a file), the `f` is confusing. The `f` means a file descriptor. It would help if you used a `path` verb.

File: station/submitobs.py [hunk 19]
```diff
@@ -56,29 +62,81 @@ def get_tle(sat_name: str, date: datetime.datetime) -> Optional[List[str]]:
...
+        _, filename = os.path.split(f)
+
+        # If there's only one file, it will use "file" key. The second file will be "file1",
+        # third "file2" etc.
+        fkey = "file" if not cnt else f"file{cnt}"
```
Comment (fivitti) [inline station/submitobs.py:132]: `cnt == 0`

File: station/submitobs.py [hunk 20]
```diff
@@ -56,29 +62,81 @@ def get_tle(sat_name: str, date: datetime.datetime) -> Optional[List[str]]:
...
+
+        file_obj = open(f, 'rb')
+        body[fkey] = file_obj
```
Comment (fivitti) [inline station/submitobs.py:135]: If you named the above variable as `file_obj`, then, maybe, it should be `file_key`?

File: station/submitobs.py [hunk 21]
```diff
@@ -56,29 +62,81 @@ def get_tle(sat_name: str, date: datetime.datetime) -> Optional[List[str]]:
...
+
+        files[fkey] = (filename, file_obj, get_mime_type(filename), {})
```
Comment (fivitti) [inline station/submitobs.py:137]: What is this empty dictionary at the end?

File: station/submitobs.py [hunk 22]
```diff
@@ -88,41 +146,45 @@ def submit_observation(data: SubmitRequestData):
...
-            file_obj,
-            "image/png",
-            {}
-        )
-    }
+    # Check if notes are a valid JSON
+    logging.debug(f"Meta-data = {data.config}")
 
-    logging.info("Submitting observation, url=%s file=%s" % (url, filename))
+    logging.info(f"Submitting observation, url={url}, file(s)={data.image_path}")
```
Comment (fivitti) [inline station/submitobs.py:152]: It looks verbose. Use the debug level.

File: station/submitobs.py [hunk 23]
```diff
@@ -88,41 +146,45 @@ def submit_observation(data: SubmitRequestData):
...
+        logging.debug("Response details: %s" % resp.text)
+        return f"Upload failed, status code {resp.status_code}"
+    else:
+        logging.info("Upload successful.")
+
+    # All seems to be OK
+    return None
 
 if __name__ == '__main__':
     if len(sys.argv) < 4:
```
Comment (fivitti) [inline station/submitobs.py:178]: It has grown a lot. We need to refactor it using the `argparse` module. But maybe in a separate issue?

**Review by fivitti — CHANGES_REQUESTED** [review]

File: station/submitobs.py (hunk 23 above)
Comment (tomaszmrugalski) [reply station/submitobs.py:178]: There's #57 created for that already.

File: server/app/controllers/receive.py (hunk 8 above)
Comment (tomaszmrugalski) [reply server/app/controllers/receive.py:65]: Updated as suggested.

File: server/app/controllers/receive.py (hunk 9 above)
Comment (tomaszmrugalski) [reply server/app/controllers/receive.py:130]: Removed DEFAULT_THUMBNAIL, using None.

File: server/app/controllers/receive.py (hunk 10 above)
Comment (tomaszmrugalski) [reply server/app/controllers/receive.py:190]: Removed.

File: server/db/svarog-11.psql (hunk 11 above)
Comment (tomaszmrugalski) [reply server/db/svarog-11.psql:15]: Hz. I've updated the code as you suggested.

File: server/templates/obslist.html (hunk 12 above)
Comment (tomaszmrugalski) [reply server/templates/obslist.html:108]: Frequencies below 1MHz are useless for satellites, but I get your point. I'm not enthusiastic: mixing MHz and GHz may mess up sorting, but I can see how it could improve readability.

File: station/submitobs.py (hunk 19 above)
Comment (tomaszmrugalski) [reply station/submitobs.py:132]: Updated as suggested.

File: station/submitobs.py (hunk 20 above)
Comment (tomaszmrugalski) [reply station/submitobs.py:135]: Renamed.

File: station/submitobs.py (hunk 21 above)
Comment (tomaszmrugalski) [reply station/submitobs.py:137]: I have no idea. This is what was there on master, before I touched it:

```
    files = {
        "file": (
            filename,
            file_obj,
            "image/png",
            {}
        )
    }
```

File: station/submitobs.py (hunk 22 above)
Comment (tomaszmrugalski) [reply station/submitobs.py:152]: This is very useful piece of information and usually there's only one file. Are you sure about debug?

**Comment by tomaszmrugalski** [comment]
Ok, last batch of changes is pushed. It adds Metadata class that maintains the metadata on station side and actually uses it for observation submissions.

File: server/templates/obslist.html (hunk 5 above)
Comment (tomaszmrugalski) [reply server/templates/obslist.html]: Since I did exactly what was requested, I suppose it's ok to resolve.

File: doc/submit_obs.md (hunk 14 above)
Comment (tomaszmrugalski) [reply doc/submit_obs.md:9]: Added as suggested.

File: station/receiver.py (hunk 15 above)
Comment (tomaszmrugalski) [reply station/receiver.py:98]: That's a good and a bit risky idea. Do you want to upload 150MB+ wav files? So maybe not all, but most. In general, I totally agree it's a great idea. Created #66 for this. It's way too large for this ticket.

File: station/submitobs.py (hunk 16 above)
Comment (tomaszmrugalski) [reply station/submitobs.py:95]: Done.

File: station/submitobs.py (hunk 17 above)
Comment (tomaszmrugalski) [reply station/submitobs.py:96]: Done.

File: station/submitobs.py (hunk 18 above)
Comment (tomaszmrugalski) [reply station/submitobs.py:127]: renamed as suggested.

**Comment by tomaszmrugalski** [comment]
@fivitti is busy with some other task that needs to be completed in 2021. I'll merge this now, but will keep the threads open, so it will be reasonably easy to get back to the discussions later, if need be.
